NFT/src/main.sw

```diff
@@ -0,0 +1,359 @@
+    /// `total_supply` has not yet been set.
+    ///
+    /// # Arguments
+    ///
+    /// * `to` - The `Identity` which will own the minted tokens.
+    /// * `amount` - The `u64` number of tokens to be minted in this transaction.
+    ///
+    /// # Reverts
+    ///
+    /// * When the sender attempts to mint more tokens than total supply.
+    /// * When the sender is not the admin and `access_control` is set.
+    #[storage(read, write)]fn mint(to: Identity, amount: u64) {
+        // The current number of tokens minted plus the amount to be minted cannot be
+        // greater than the total supply
+        require(storage.token_supply >= (storage.token_count + amount), InputError::NotEnoughTokensToMint);
+
+        // Ensure that the sender is the admin if this is a controlled access mint
+        require(!storage.access_control || (storage.admin.is_some() && msg_sender().unwrap() == storage.admin.unwrap()), AccessError::SenderDoesNotHaveAccessControl);
+
+        // Mint as many tokens as the sender has asked for
+        let mut index = 0;
+        let mut minted_tokens = ~Vec::new::<u64>();
+        while index < amount {
+            // Increment the token count
+            storage.token_count = storage.token_count + 1;
+
+            // Create the metadata for this new token with the owner
+            let meta_data = ~MetaData::new(Option::None(), to);
+            storage.meta_data.insert(storage.token_count, Option::Some(meta_data));
+
+            // Increase the balance of the new owner
+            storage.balances.insert(to, storage.balances.get(to) + 1);
+
+            // Push to minted tokens Vec
+            minted_tokens.push(storage.token_count);
+
+            // Increment the number of tokens minted in this transaction
+            index = index + 1;
+        }
+
+        log(MintEvent {
+            owner: to, token_ids: minted_tokens
+        });
+    }
+
+    /// Burns the specified token. When burned, the NFT Metadata of the token is set
+    /// to `None`. After the token has been burned, no one will be able to fetch any data
+    /// about this token or have control over it.
+    ///
+    /// # Arguments
+    ///
+    /// * `token_id` - The `u64` ID of the token which is to be burned.
+    ///
+    /// * Reverts
+    ///
+    /// * When `token_id` does map to an existing token.
+    /// * When sender is not the owner of the `token_id`.
+    #[storage(read, write)]fn burn(token_id: u64) {
+        // Ensure this is a valid token that has already been minted and exists
+        let mut meta_data = token_metadata(storage.meta_data.get(token_id));
+
+        // Ensure the sender owns the token that is provided
+        let owner = msg_sender().unwrap();
+        require(meta_data.owner == owner, AccessError::SenderNotOwner);
+
+        // Burn this token by setting the `token_id` Metadata mapping to `None`
+        storage.meta_data.insert(token_id, Option::None());
```

Comment:
I'm not sure here about the gas cost implications of setting the metadata to `None`...
Are we effectively clearing (overwriting with 0's) all slots used by a Metadata struct (up to 4 slots here, 2 for each `Identity`)? 
If so, then we should use a different approach for registering a token as burned (i.e: keep `owner` separate so that just the owner can be set to `None`.
cc @mohammadfawaz

Reply:
@nfurfaro is right. We're not effectively overwriting with zeros thought because `None` is not the 0th enum variant, but we still write `size_of(Option<MetaData>)` bytes which you can divide by 32 and round up to figure out how many storage slots it requires.

Reply:
I did a quick test for this, which indicates that we're writing 3 slots for TokenMetadata:
```
script;

use std::{
    address::Address,
    identity::Identity,
    intrinsics::size_of_val,
    logging::log,
    option::Option,
};

pub struct TokenMetaData {
    /// The user that is allowed to transfer this token.
    approved: Option<Identity>,
    /// The user that owns this token.
    owner: Identity,
}

fn main() -> u64 {

    let address_1 = ~Address::from(0x1111111111111111111111111111111111111111111111111111111111111111);
    let approved_addr = Identity::Address(address_1);
    let approved_opt = Option::Some(approved_addr);
    let address_2 = ~Address::from(0x2222222222222222222222222222222222222222222222222222222222222222);
    let owner = Identity::Address(address_2);

    let metadata = TokenMetaData {
        approved: approved_opt,
        owner: owner,
    };

    let size = size_of_val(metadata);
    let div = size / 32;
    let mod = size % 32;

    let total_slots = if mod > 0 {
        div + 1
    } else {
        div
    };

    total_slots
}
```

Reply:
I suggest breaking up the `TokenMetadata` into 2 separate storage variables which can be updated independently of each other, as the organizational benefit of grouping them is eclipsed by the extra gas cost while managing them.

Reply:
What is the original gas cost vs your proposed gas cost i.e. the difference in cost?

Reply:
I can't speak to the actual cost, but we would eliminate reading/writing 1 slot with the proposed change (so ~33% savings), **each time** we need to read or write **either** of the 2 TokenMetaData fields.

(The change proposed [here](https://github.com/FuelLabs/fuel-specs/issues/324#issuecomment-1146847164)
would eliminate another storage slot, but is orthogonal to this PR.)

Reply:
I've personally never been a fan of the way Solidity NFTs use a URI and store data off-chain, but I understand the reasoning. I'd like to limit that as much as possible. As for these two methods of storing the URI, I defintely prefer the URI-shortener to have a uniform length. I will have to dig deeper into this tomorrow or any other possible solutions.

Reply:
Can we defer this to a future PR? Interesting discussion, but this level of gas golfing isn't a req for functionality.